org.eclipse.jdt.core.tests.model/src/org/eclipse/jdt/core/tests/model/ReconcilerTests16.java

```diff
@@ -299,4 +299,44 @@ public void testissue342_001() throws Exception {
 		deleteProject(p);
 	}
 }
+public void testIssue612_001() throws Exception {
+	if (!isJRE16)
+		return;
+	IJavaProject p = createJava16Project("p");
+	createFolder("/p/src/a");
+	try {
+		String source =
+				"package a;\n"
+				+ "public class X {\n"
+				+ "    public static void main(String[] args) {\n"
+				+ "        Sub sub = new Sub();\n"
+				+ "        sub.method1();\n"
+				+ "    }\n"
+				+ "\n"
+				+ "    static class Outer<T> { \n"
+				+ "        class Inner {\n"
+				+ "            void sayHi() {\n"
+				+ "                System.out.println(\"hello world!\");\n"
+				+ "            }\n"
+				+ "        }\n"
+				+ "    }\n"
+				+ "\n"
+				+ "    static class Sub extends Outer<Sub.Inner> {\n"
```

Comment:
I can break your fix by changing `extends` to `implements` with the necessary changes to `Outer`.

Reply:
Fixed.